src/lib/reasoners/intervalCalculus.ml

```diff
@@ -2045,17 +2045,35 @@ let case_split env uf ~for_model =
       end
   | _ -> res
 
-let optimizing_split env uf opt_split =
+(* Helper function used in [optimizing_objective] to pick a value
+   for the polynomial [p] in its interval. We used this function in the case
+   the value produced by the optimization procedure doesn't satisfy some
+   constraints that involve strict inequalities or the problem is unbounded. *)
+let middle_value env ~to_max ty p bound =
+  let interval =
+    match MP0.find_opt p env.polynomes, bound with
+    | Some i, Some bound ->
+      begin
+        try
+          if to_max then
+            Intervals.new_borne_sup Ex.empty bound ~is_le:false i
+          else
+            Intervals.new_borne_inf Ex.empty bound ~is_le:false i
+        with Intervals.NotConsistent _ -> assert false
+      end
+    | Some i, None -> i
+    | None, _ -> Intervals.point Q.zero ty Ex.empty
+  in
+  let q = Option.get (Intervals.pick ~to_max interval) in
+  alien_of (P.create [] q ty)
+
+let optimizing_objective env uf Objective.Function.{ e; to_max; _ } =
   (* soundness: if there are expressions to optmize, this should be
      done without waiting for ~for_model flag to be true *)
-  let {Th_util.r = r; is_max = to_max; e; value; _ } = opt_split in
-  assert (match value with
-      | Unknown -> true
-      | _ -> false
-    );
+  let uf, _ = Uf.add uf e in
```

Comment:
Uhm, I don't think we need to keep the representative element here. The expression `e` here isn't a part of our problem but the expression we try to optimize modulo our constraints. Now, I remember why I add this line before 
```ocaml 
let repr, _ = Uf.find uf e in
```
 
The reason is because the function `Uf.find` fails if the expression have never been added to the union-find before. As the expression can be something completely new, this can happen. Consider for instance the input file:
```
(set-logic ALL)
(set-option :produce-models true)
(declare-const x Int)
(assert (<= x 10))
(maximize (+ x 5))
(check-sat)
(get-model)
(get-objectives)
```
The expression `(+ x 5)` is completely new and the union-find will raise `Not_found` exception here.

Now if the expression is already knew, we want to get both its representative in the union-find (in the code it's `repr`) and the semantic values returned by `X.make` for this expression (in the code it's `r1`).

Another solution would be add the expression `e` in the union-find when we call the function `optimize` of the SAT solver.

Reply:
If we want to get the representative if there is one or the result of `X.make` otherwise then the right code is probably along the lines of:

```ocaml
let repr, r1 =
  match Uf.find uf e with
  | repr, _ -> repr, Uf.make uf e
  | exception Not_found -> let r, _ = X.make e in r, r
```

That said, I didn't think about it, but you are right — we should add `e` in the call to `optimize` (this is what was done before when `Optimize` was a symbol). Note that this needs to use `Theory.add_term` to make sure that all sub-terms are properly added (`Uf.add` only adds the term itself but not the subterms).

The real reason to do this is that if we do `(maximize (bv2nat x))` for instance, and the problem does not otherwise contain `(bv2nat x)`, not adding `bv2nat` to the theory would lose some reasoning power (because it would not be seen by the relations, and knowledge about it would not be properly propagated).